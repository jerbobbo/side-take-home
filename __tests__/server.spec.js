const { ApolloServer } = require('apollo-server-express');
const { Listings } = require('../stores/mongodb/Listings');
const { getListingsCollection } = require('../stores/mongodb/store');
const { SimplyRetsAPI } = require('../stores/simplyrets/simply-rets-api');
const typeDefs = require('../types');
const resolvers = require('../resolvers');
const { getListingsMock } = require('./resources/simplyRetsMocks');
const { findOrCreateMock, addToFavoriteCountMock } = require('./resources/listingsMocks');
const { getListingsQuery, markAsFavoriteMutation } = require('./resources/queries');

describe('Test Apollo Server', () => {
  let simplyRetsApi;
  let listings;
  let server;

  beforeEach(() => {
    simplyRetsApi = new SimplyRetsAPI();
    listings = new Listings({ store: getListingsCollection() });

    // create test server using our typeDefs, resolvers, dataSources
    server = new ApolloServer({
      typeDefs,
      resolvers,
      dataSources: () => ({
        simplyRetsApi,
        listings,
      }),
    });
  });
  
  it('Gets listings by city', async () => {
    // mock the dataSource's underlying find methods
    const getListingsSpy = jest
      .spyOn(simplyRetsApi, 'getListings')
      .mockResolvedValue([getListingsMock]);

    const findOrCreateListingSpy = jest
      .spyOn(listings, 'findOrCreateListingById')
      .mockResolvedValue(findOrCreateMock);

    // run query against the server and snapshot the output
    const res = await server.executeOperation({
      query: getListingsQuery,
    });

    expect(getListingsSpy).toBeCalledWith('Houston');
    const expectedListingId = getListingsMock.listingId;
    expect(findOrCreateListingSpy).toBeCalledWith(expectedListingId);
    expect(res).toMatchSnapshot();
  });

  it('Increments favorite count with markAsFavorite mutation', async () => {

    // mock the addToFavoriteCount method
    const addToFavoriteCountSpy = jest
      .spyOn(listings, 'addToFavoriteCount')
      .mockResolvedValue(addToFavoriteCountMock);

    // run query against the server and snapshot the output
    const res = await server.executeOperation({
      query: markAsFavoriteMutation,
    });

    expect(addToFavoriteCountSpy).toBeCalledWith('49699701');
    expect(res).toMatchSnapshot();
  });

});
