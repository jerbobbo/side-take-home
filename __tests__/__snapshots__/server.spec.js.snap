// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Apollo Server Gets listings by city 1`] = `
Object {
  "data": Object {
    "properties": Array [
      Object {
        "address": Object {
          "city": "Tomball",
          "country": "United States",
          "crossStreet": "High St / Main St. E",
          "full": "32738 South VISTA MADERA Lane #F16",
          "postalCode": "77024",
          "state": "Texas",
          "streetName": "South VISTA MADERA Lane",
          "streetNumber": 32738,
          "streetNumberText": "32738",
          "unit": "F16",
        },
        "disclaimer": "This information is believed to be accurate, but without warranty.",
        "favoriteCount": 0,
        "listPrice": 2893021,
        "listingId": "6323997",
        "property": Object {
          "area": 1562,
          "bedrooms": 2,
        },
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Test Apollo Server Increments favorite count with markAsFavorite mutation 1`] = `
Object {
  "data": Object {
    "markAsFavorite": Object {
      "favoriteCount": 1,
      "message": "Listing 49699701 successfully marked as favorite.",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
